fastavro/_validate_py.py

```diff
@@ -0,0 +1,357 @@
+    """
+    Check that the data is a Mapping type with all schema defined fields
+    validated as True.
+
+    :param datum: Mapping: data to validate schema fields
+    :param schema: avro schema of 'record' type
+    :param parent_ns: str: parent namespace
+    :param raise_errors: bool: should raise ValidationError
+    :return: bool
+    :except: ValidationError
+    """
+    if raise_errors:
+        namespace, name = schema_name(schema, parent_ns)
+    else:
+        name = parent_ns
+    return (
+        isinstance(datum, Mapping) and
+        all(validate(datum=datum.get(f['name'], f.get('default')),
+                     schema=f['type'],
+                     field=schema_name(f, name)[1] if raise_errors else name,
+                     raise_errors=raise_errors)
+            for f in schema['fields']
+            )
+    )
+
+
+def validate_union(datum, schema, parent_ns=None, raise_errors=False):
+    """
+    Check that the data is a list type with possible options to
+    validate as True.
+
+    :param datum: (Iterable, tuple(name, Iterable)): data to validate
+    as multiple data types
+    :param schema: avro schema of 'union' type
+    :param parent_ns: str: parent namespace
+    :param raise_errors: bool: should raise ValidationError
+    :return: bool
+    :except: ValidationError
+    """
+    if isinstance(datum, tuple):
+        (name, datum) = datum
+        for candidate in schema:
+            if extract_record_type(candidate) == 'record':
+                if name == candidate["name"]:
+                    return validate(datum, schema=candidate,
+                                    field=parent_ns,
+                                    raise_errors=raise_errors)
+        else:
+            return False
+
+    errors = []
+    for s in schema:
+        try:
+            ret = validate(datum, schema=s,
+                           field=parent_ns,
+                           raise_errors=raise_errors)
+            if ret:
+                # We exit on the first passing type in Unions
+                return True
+        except ValidationError as e:
+            errors.extend(e.errors)
+    if raise_errors:
+        raise ValidationError(*errors)
+    return False
+
+
+BASE_VALIDATORS = {
+    'null': validate_null,
+    'boolean': validate_boolean,
+    'string': validate_string,
+    'int': validate_int,
+    'long': validate_long,
+    'float': validate_float,
+    'double': validate_float,
+    'bytes': validate_bytes,
+    'fixed': validate_fixed,
+    'enum': validate_enum,
+    'array': validate_array,
+    'map': validate_map,
+    'union': validate_union,
+    'error_union': validate_union,
+    'record': validate_record,
+    'error': validate_record,
+    'request': validate_record
+}
+
+VALIDATORS = BASE_VALIDATORS.copy()
+
+
+def register_validator(record_type, validator):
+    if record_type in BASE_VALIDATORS:
+        raise ValueError("Not allowed to override Base Validators.")
```

Comment:
The only reasoning I could think of, is if we wanted to support numpy types for int,long, it would easy for someone to just register on top of it until we get an actual pull request to update the original.

Reply:
Technically updates to any types in BASE_VALIDATORS will only work for the python version, since the validate function calls the functions directly for speed, if we want to lose the speed, we can have the `cpdef validate` use `get_validator` instead of the large if conditions.